### PR TITLE
global: major rewrite

### PR DESCRIPTION
**Pull request by jirikuncar**
Signed-off-by: Jiri Kuncar jiri.kuncar@cern.ch
- [x] add more examples to docs and `examples/app.py`
## 

cc @egabancho @tiborsimko @lnielsen

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -22,40 +22,63 @@
...
+    'Babel>=1.3',
+]
+
+install_requires = [
+    'Flask-BabelEx>=0.9.2',
+    'Flask-CLI>=0.2.1',
+    'fs>=0.5.3',
+    'invenio-db>=1.0.0a4',
+    'invenio-records>=1.0.0a2',
+    'jsonpointer>=1.9',
```
Comment (tiborsimko) [inline setup.py:74]: As discussed IRL, it's OK to depend on `invenio-records`, though for later use cases (e.g. CDS commenting system that is accepting full-text document files) we may perhaps revive the notion of standalone documents.  

(This was originally introduced for INSPIRE plots's needs, but I think this is no longer required in the optics of "everybody is a first class citizen" files included.)

**Comment by tiborsimko** [comment]
Can you please amend the order of import statements to make Travis happy?

```
 find . -name "*.py" -exec isort {} \;
```

Otherwise everything works well for me locally...

File: AUTHORS.rst [hunk 2]
```diff
@@ -21,16 +21,10 @@
     waive the privileges and immunities granted to it by virtue of its status
```
Comment (lnielsen) [inline AUTHORS.rst:21]: Keep old authors?

File: tests/test_invenio_documents.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+        assert os.path.exists(hello_strpath)
+
+        with open(bye_strpath, 'wb') as f:
+            f.write(u'Bye bye'.encode('utf-8'))
+
+        with app.app_context():
+            record = Record.create({'title': 'Greetings',
+                                    'document': hello_strpath})
+            record_id = record.model.id
```
Comment (lnielsen) [inline tests/test_invenio_documents.py:140]: Assigned but not used

File: invenio_documents/api.py [hunk 4]
```diff
@@ -1,231 +1,105 @@
...
 
-            >>> from flask import g
-            >>> from invenio_base.factory import create_app
-            >>> app = create_app()
-            >>> ctx = app.test_request_context()
-            >>> ctx.push()
-            >>> from invenio_document import api
-            >>> d = api.Document.create({'title': 'Title 1'})
-            >>> e = api.Document.get_document(d['_id'])
+import fs
```
Comment (lnielsen) [inline invenio_documents/api.py]: Unused import

File: invenio_documents/cli.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+from __future__ import absolute_import, print_function
+
+import json
+import sys
+
+import click
+from flask_cli import with_appcontext
+
+from invenio_db import db
```
Comment (lnielsen) [inline invenio_documents/cli.py]: Unused import

File: tests/test_invenio_documents.py (hunk 3 above)
Comment (jirikuncar) [inline tests/test_invenio_documents.py:140]: true, I have forgot to put it https://github.com/inveniosoftware/invenio-documents/pull/9/files#diff-841047e6c91809a0f6db09e83439fcb1R143 and https://github.com/inveniosoftware/invenio-documents/pull/9/files#diff-841047e6c91809a0f6db09e83439fcb1R151

**Comment by jirikuncar** [comment]
> Can you please amend the order of import statements to make Travis happy?

@tiborsimko I did it several times (see Travis CI build history). There are some issues with wrong detection of `invenio_*` packages.

File: invenio_documents/cli.py (hunk 5 above)
Comment (jirikuncar) [inline invenio_documents/cli.py]: :+1:

File: invenio_documents/api.py (hunk 4 above)
Comment (jirikuncar) [inline invenio_documents/api.py]: :+1: (late refactoring :wink: )

File: AUTHORS.rst (hunk 2 above)
Comment (jirikuncar) [inline AUTHORS.rst:21]: sure - I missed that

**Comment by lnielsen** [comment]
:shipit: 

I'm not sure about if the CLI is too low-level (i.e. giving too much power to mess things up), but let's defer this discussion until we have looked into integrating documents with records-rest.

**Comment by tiborsimko** [comment]
> There are some issues with wrong detection of invenio_\* packages.

Yes, I saw these before, sometimes even due to different Python 3.4 version on Travis CI and locally.

FWIW, on my local box, these files need to be touched:

```
$ git status  | grep modified
        modified:   examples/app.py
        modified:   invenio_documents/cli.py
        modified:   tests/conftest.py
        modified:   tests/test_invenio_documents.py
```

Perhaps you can simply add `# isort:skip` to incriminated import statements for now, so that these cosmetic comments wouldn't over-shade the important stuff in the PR red/greed light status (=the tests).

(or even `isort:skip_file`...)

**Comment by egabancho** [comment]
:shipit:

**Comment by jirikuncar** [comment]
> I'm not sure about if the CLI is too low-level (i.e. giving too much power to mess things up)

@lnielsen Yes, you can a control over everything, however it's **only** in CLI and not in REST interface.

**Comment by lnielsen** [comment]
> @lnielsen Yes, you can a control over everything, however it's only in CLI and not in REST interface.

Yes I understand. I was just musing over if these operations might even be too much power to enable in a CLI.

Anyway, once more parts of Invenio 3 has been done, we can review what makes sense. E.g. we also need to see where we include stuff such as checksum computation of files etc (probably not in this module but let's see).

**Comment by jirikuncar** [comment]
I can add `--yes-i-know` to show that the commands are :exclamation: dangerous :exclamation:

**Comment by jirikuncar** [comment]
@lnielsen can we `:shipit:`?

**Comment by lnielsen** [comment]
Yes, for sure :-) Already put the :shipit: :-)

**Pull request merged**